.ci/pitest.groovy

```diff
@@ -0,0 +1,322 @@
+    }
+    return survivedMutations
+}
+
+private static Set<Mutation> getSuppressedMutationList(Node mainNode) {
+    def children = mainNode.children()
+    Set<Mutation> suppressedMutations = new TreeSet<>()
+
+    children.each { node ->
+        Node mutationNode = node as Node
+        suppressedMutations.add(getMutation(mutationNode))
+    }
+    return suppressedMutations
+}
+
+private static Mutation getMutation(Node mutationNode) {
+    List childNodes = mutationNode.children()
+
+    String sourceFile = null
+    String mutatedClass = null
+    String mutatedMethod = null
+    String mutator = null
+    String lineContent = null
+    String description = null
+    String mutationClassPackage = null
+    int lineNumber = 0
+    childNodes.each {
+        Node childNode = it as Node
+        String text = childNode.name()
+
+        String childNodeText = XmlUtil.escapeXml(childNode.text())
+        switch (text) {
+            case "sourceFile":
+                sourceFile = childNodeText
+                break
+            case "mutatedClass":
+                mutatedClass = childNodeText
+                mutationClassPackage = mutatedClass.split("[A-Z]")[0]
+                break
+            case "mutatedMethod":
+                mutatedMethod = childNodeText
+                break
+            case "mutator":
+                mutator = childNodeText
+                break
+            case "description":
+                description = childNodeText
+                break
+            case "lineNumber":
+                lineNumber = Integer.parseInt(childNodeText)
+                break
+            case "lineContent":
+                lineContent = childNodeText
+                break
+        }
+    }
+    if (lineContent == null) {
+        String mutationFileName = mutationClassPackage + sourceFile
+        String startingPath = ".${SEPARATOR}src${SEPARATOR}main${SEPARATOR}java${SEPARATOR}"
+        String javaExtension = ".java"
+        mutationFileName = startingPath + mutationFileName
+                .substring(0, mutationFileName.length() - javaExtension.length())
+                .replaceAll("\\.", SEPARATOR) + javaExtension
+
+        File file = new File(mutationFileName)
+        lineContent = XmlUtil.escapeXml(file.readLines().get(lineNumber - 1).trim())
+    }
+    if (lineNumber == 0) {
+        lineNumber = -1
+    }
+    boolean unstable = false
+    String unstableMutation = mutationNode.attribute("unstable")
+    if (unstableMutation != null) {
+        unstable = Boolean.parseBoolean(unstableMutation)
+    }
+    return new Mutation(sourceFile, mutatedClass, mutatedMethod, mutator, description,
+            lineContent, lineNumber, unstable)
+}
+
+private static int compareSurvivedAndSuppressedMutations(Set<Mutation> survivedMutations,
+                                                         Set<Mutation> suppressedMutations) {
+
+    int exitCode = 0
+    if (survivedMutations != suppressedMutations) {
```

Comment:
General comment here: can we declare the differences of sets at the beginning of this method, and make conditions/ branches simpler here, with one return point and one if/else chain?

What exactly are our conditions? Please put them in Javadoc.

1. There are mutations in report that are not in suppressions -> fail
2. There are mutations in suppressions that are not in report, but they are all unstable -> pass (this condition could be boolean extracted as set difference -> stream -> findFirst)
3. There are mutations in suppressions that are not in report -> fail
4. Report and suppressions are equal -> pass

Reply:
> one return point and one if/else chain?

One `if-else` chain is not possible as we need to check both `survived - suppressed` and `suppressed - survived`.

Added conditions in javadoc.

Reply:
> survived - suppressed and suppressed - survived

Can we do:

```
final int exitcode;
if (survived-suppressed is not empty) {
    exitcode = 1
    //print stuff
}
else if (suppressed - survived is all unstable) {
    exitcode = 0
    // print stuff
}
else if (suppressed - survived is not empty) {
    exitcode = 1
    // print stuff
}
else {
    exitcode = 0
    // print stuff
}

```
?

Reply:
This cannot be done as multiple branches can be true at the same time->
```
survived - suppressed is not empty -> true, survived mutations 
```
and 
```
suppressed - survived is not empty -> true, unnecessary suppressions.
```
Both these branches can be true at the same time for any profile, so refactoring to one `if-else` chain is not possible.

Reply:
> Both these branches can be true at the same time for any profile

Good point. Then let's go with previous script's behavior:

```
int exitcode;
if (no diff) {
    exitcode = 0
}
else if (diff is all unstable) {
    exitcode = 0
}
else {
    // we know diff should fail execution
    // show diff
    exitcode = 1
}

```

Reply:
Done.

Reply:
@nick-mancuso Please see the latest changes, since now we are not returning anything now (`System.exit` is not possible, CodeNarc code smell), the Javadoc comment is still present, if you insist I can add empty if branches with comments like `// do nothing`.

Reply:
Please suppress this rule, we can refine script later if we need to.

Reply:
Done, added suppression.